### PR TITLE
Add py-clang-format

### PR DESCRIPTION
**Pull request by berquist**
This package is for a PyPI-provided standalone `clang-format` that normally requires all of LLVM.  I need 12.0.1, hence all the old versions; I've confirmed 10.0.1, 12.0.1, and 17.0.6 have the right hashes and are working.

Since it is using the source distribution and not one of the wheels, it uses CMake to fetch all of LLVM at the target version and compiles just the `clang-format` target.  It's slow, uses a lot of disk space, and there are probably hidden dependencies I'm missing.  I'd like to use wheels instead but I'm not sure of the best way.  Is it like [py-itk](https://github.com/spack/spack/blob/f27bff81ba9cd53416ceda109aa9a6708dd13760/var/spack/repos/builtin/packages/py-itk/package.py)?

### CODE REVIEW

**Review by tldahlgren — COMMENTED** [review]
FYI. Confirmed the homepage and all 36 version sha256s.

**Comment by adamjstewart** [comment]
We prefer to not use wheels. As you can tell from py-itk, it's kind of a nightmare.

We also prefer not to download dependencies from the internet. Is it possible for this package to depend on LLVM, or to fetch LLVM as a resource?

**Comment by berquist** [comment]
> We prefer to not use wheels. As you can tell from py-itk, it's kind of a nightmare.

Yeah, I didn't want to have to do that.

> We also prefer not to download dependencies from the internet. Is it possible for this package to depend on LLVM, or to fetch LLVM as a resource?

Yes, I can use a patch to rewrite their CMakeLists.txt for this.  I almost think it's not worth it for this package to exist just to have LLVM as a build dependency, but a copy of v16 is 16 gigabytes on this machine...so it is useful for that alone.

---

Unrelated, it seems like I am no longer able to convert PRs here to draft status.

File: var/spack/repos/builtin/packages/py-clang-format/package.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+    maintainers("berquist")
+
+    license("Apache-2.0", checked_by="berquist")
+
+    for ver, sha256 in _versions:
+        version(ver, sha256=sha256)
+        depends_on("llvm@={} +clang".format(ver), when="@={}".format(ver), type="build")
+
+    patch("use_local_llvm.patch")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-clang-format/package.py]: Does this patch work for every version? Ideally, we could write a patch that would be possible to upstream so we don't have to keep changing it every time the format changes.

**Comment by berquist** [comment]
Here are version-range patches.  I can make a change upstream but that won't handle older versions.

File: var/spack/repos/builtin/packages/py-clang-format/use_local_llvm.patch
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-clang-format/use_local_llvm.patch:1]: This one is no longer used?

**Comment by adamjstewart** [comment]
> I can make a change upstream but that won't handle older versions.

Please do. I don't want to keep making patches for future versions and want to get feedback from the developers.

**Comment by berquist** [comment]
I don't have the time to finish this properly and don't need to use `clang-format` through Spack.

**Pull request closed without merging**